docs/guides/high-availability/concepts-memory-and-cpu-sizing.adoc

```diff
@@ -0,0 +1,96 @@
+
+* Use these values as a starting point and perform your own load tests before going into production.
+====
+
+Summary:
+
+* The used CPU scales linearly with the number of requests up to the tested limit below.
+* The used memory scales linearly with the number of active sessions up to the tested limit below.
+
+Recommendations:
+
+* The base memory usage for an inactive Pod is 1 GB of RAM.
+
+* Leave 1 GB extra head-room for spikes of RAM.
+
+* For each 100,000 active user sessions, add 500 MB per Pod in a three-node cluster (tested with up to 200,000 sessions).
++
+This assumes that each user connects to only one client.
+Memory requirements increase with the number of client sessions per user session (not tested yet).
+
+* For each 40 user logins per second, 1 vCPU per Pod in a three-node cluster (tested with up to 300 per second).
++
+{project_name} spends most of the CPU time hashing the password provided by the user.
+
+* For each 450 client credential grants per second, 1 vCPU per Pod in a three node cluster (tested with up to 2000 per second).
++
+Most CPU time goes into creating new TLS connections, as each client runs only a single request.
+
+* For each 350 refresh token requests per second, 1 vCPU per Pod in a three node cluster (tested with up to 435 refresh token requests per second).
+
+* Leave 200% extra head-room for CPU usage to handle spikes in the load.
+This ensures a fast startup of the node, and sufficient capacity to handle failover tasks like, for example, re-balancing Infinispan caches, when one node fails.
+Performance of {project_name} dropped significantly when its Pods were throttled in our tests.
+
+=== Calculation example
+
+Target size:
+
+* 50,000 active user sessions
+* 40 logins per seconds
+* 450 client credential grants per second
+* 350 refresh token requests per second
+
+Limits calculated:
+
+* CPU requested: 3 vCPU
++
+(40 logins per second = 1 vCPU, 450 client credential grants per second = 1 vCPU, 350 refresh token = 1 vCPU)
+
+* CPU limit: 9 vCPU
++
+(Allow for three times the CPU requested to handle peaks, startups and failover tasks, and also refresh token handling which we don't have numbers on, yet)
+
+* Memory requested: 1.25 GB
++
+(1 GB base memory plus 250 MB RAM for 50,000 active sessions)
+
+* Memory limit: 2.25 GB
++
+(adding 1 GB to the memory requested)
+
+== Reference architecture
+
+The following setup was used to retrieve the settings above to run tests of about 10 minutes for different scenarios:
+
+* OpenShift 4.13.x deployed on AWS via ROSA.
+* Machinepool with `m5.4xlarge` instances.
+* {project_name} deployed with the Operator and 3 pods.
+* Default user password hashing with PBKDF2 27,500 hash iterations.
+* Database seeded with 100,000 users and 100,000 clients.
+* Infinispan caches at default of 10,000 entries, so not all clients and users fit into the cache, and some requests will need to fetch the data from the database.
```

Comment:
As stated here 10k max-entries is not enough to store all realm data. Based on the 100k users/clients don't we want to recommend a better limit for local caches? I don't know how to calculate this but ideally, we should keep the state in memory as much as possible to avoid unnecessary database roundtrips?

Reply:
In a real world scenario you wouldn't have all the users in memory anyway, and especially on restarting nodes those would have an empty cache at the beginning. 

Running a load test where Keycloak is loading data from the database continuously and succeeds gives me more confidence than a load test where all data is in memory all the time. The 10_000 cache entries are enough to ensure that a user is cached during the authentication flow, which is good enough and already reduces a lot of the load.

Reply:
I'm not 100% convinced about that based on the deployments I've seen. This is especially true when you have a load spike and failures that force users to re-authenticate.

For the realm `cache`, depending on the number of roles, groups, clients, and other related data, this number is also not enough.

For production deployments I don't think the right recommendation should be use the default values. But at least recommend trying out increasing these local caches to match the expected performance and database resource usage (IIOPS are expensive depending on your deployment and we had cases around that in the past).

Don't you think we should at least mention the points we are discussing here? For instance:

```
The default values for local caches such as `realm` and `users` caches should be adjusted accordingly to your performance requirements, database resource usage, and memory constraints. Make sure to run load tests as close as possible to your real load and monitor how resources are being used to better adjust these settings. 
```

Reply:
Btw. the local caches (https://github.com/keycloak/keycloak/blob/main/quarkus/runtime/src/main/resources/cache-ispn.xml#L26) in don't seem to have any expiration policy configured. I think it would be helpful to enable some expiration to allow keycloak to release more memory if the user isn't used for a while (e.g. 8h).

Reply:
> The local caches don't seem to have any expiration policy configured.

@thomasdarimont - this is not related to this guide, and probably a default that Keycloak should change. Please create an issue in Keycloak's main repository and submit a pull request with your ideas around that and discuss it. It could be 8 or 24 or 36 hours. Ping me and @pedroigor in the PR and we'll listen to your thoughts on this.

Reply:
> For production deployments I don't think the right recommendation should be use the default values. But at least recommend trying out increasing these local caches to match the expected performance and database resource usage (IIOPS are expensive depending on your deployment and we had cases around that in the past).

There is already a paragraph at the top with recommendations to increase the cache. Will add a hit towards IOPS there in https://github.com/keycloak/keycloak/issues/24975.

I think this is best tuned when looking at the metrics. For https://github.com/keycloak/keycloak-benchmark/issues/579, I've added "Cache hit ratio + cache size for local caches".